### PR TITLE
Fixes #811

### LINKED ISSUE #811: rooted atom fingerprint non identical for the same molecules
When I calculate the rooted atom torsion fingerprint for various molecules that should be identical the fingerprints are not identical. Here is a simplified example

```python
m0 = Chem.MolFromMolBlock("""C6H6BN3
     RDKit          3D

 10 11  0  0  0  0  0  0  0  0999 V2000
   -0.0000   -0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4000   -0.0000    0.0000 B   0  0  0  0  0  0  0  0  0  0  0  0
    2.9037   -3.0062    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.5722   -3.4388    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.5318   -2.5020    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    3.1948   -1.6368    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.1544   -0.7000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    0.8229   -1.1326    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.8229    1.1326    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.1544    0.7000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  9 10  2  0
 10  7  1  0
  7  6  1  0
  7  8  1  0
  3  4  1  0
  3  6  2  0
  4  5  2  0
  5  8  1  0
  8  1  2  0
  1  9  1  0
M  END
""")

m1 = Chem.MolFromMolBlock("""C6H6BN3
     RDKit          2D

 10 11  0  0  0  0  0  0  0  0999 V2000
   -0.0000   -0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4000   -0.0000    0.0000 B   0  0  0  0  0  0  0  0  0  0  0  0
    0.8229    1.1326    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.1544    0.7000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    2.1544   -0.7000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    3.1948   -1.6368    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.9037   -3.0062    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.5722   -3.4388    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.5318   -2.5020    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    0.8229   -1.1326    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0
  1  3  1  0
  3  4  2  0
  4  5  1  0
  6  7  2  0
  7  8  1  0
  8  9  2  0
  5  6  1  0
 10  1  2  0
 10  9  1  0
  5 10  1  0
M  END
""")

In [5]: print Chem.MolToSmiles(m0), Chem.MolToSmiles(m1)
Bc1cnn2cccnc12 Bc1cnn2cccnc12
```
When I now do a rooted fingerprint based on the boron atom I get:

```python
BQuery = Chem.MolFromSmarts('B')
match0 = m0.GetSubstructMatch(BQuery)
nfp0 = Torsions.GetHashedTopologicalTorsionFingerprint(m0,nBits=9192,targetSize=7,fromAtoms=match0)
    
match1 = m1.GetSubstructMatch(BQuery)
nfp1 = Torsions.GetHashedTopologicalTorsionFingerprint(m1,nBits=9192,targetSize=7,fromAtoms=match1)
print nfp0.GetNonzeroElements(), nfp1.GetNonzeroElements()
{6040: 1, 8805: 1, 8373: 1, 8182: 1, 8087: 1} {6040: 1, 8151: 1, 8373: 1, 8182: 1, 8087: 1}
```
As one can see the positions 8151 and 8805 do not match between the molecules

When the molecules are canonicalised differently things look better:

```python
rank1 = Chem.CanonicalRankAtoms(m0)
m0 = Chem.RenumberAtoms(m0, rank1)
rank1 = Chem.CanonicalRankAtoms(m1)
m1 = Chem.RenumberAtoms(m1, rank1)
​
match0 = m0.GetSubstructMatch(BQuery)
match1 = m1.GetSubstructMatch(BQuery)
nfp0 = Torsions.GetHashedTopologicalTorsionFingerprint(m0,nBits=9192,targetSize=7,fromAtoms=match0)
nfp1 = Torsions.GetHashedTopologicalTorsionFingerprint(m1,nBits=9192,targetSize=7,fromAtoms=match1)
a1 = nfp0.GetNonzeroElements()
a2 = nfp1.GetNonzeroElements() 
a1 == a2
Out[77]:
True
```
This time also the FPs look better:

```python
print a1, a2
{6040: 1, 8805: 1, 8373: 1, 8182: 1, 8087: 1} {6040: 1, 8805: 1, 8373: 1, 8182: 1, 8087: 1}
```
but it is nothing I would think that I need to do.
Thx
NIk

### PR DESCRIPTION
**Pull request by greglandrum**
more detail is in the comments of #811.

### CODE REVIEW

File: Code/GraphMol/Fingerprints/AtomPairs.cpp [hunk 1]
```diff
@@ -438,17 +438,17 @@ SparseIntVect<boost::int64_t> *getTopologicalTorsionFingerprint(
...
            ++pIt) {
         // look for a cycle that doesn't start at the first atom
         // we can't effectively canonicalize these at the moment
         // (was github #811)
-        if (pIt != path.begin()) {
-          if (std::find(pIt + 1, path.end(), *pIt) != path.end()) {
-            pathCodes.clear();
-            break;
-          }
+        if (pIt != path.begin() && pAtoms[*pIt]) {
```
Comment (bp-kelley) [inline Code/GraphMol/Fingerprints/AtomPairs.cpp]: Ah.  Nice!

**Comment by greglandrum** [comment]
crap. going to have to track down that test failure. I'll fix that later.

File: Code/GraphMol/Fingerprints/AtomPairs.cpp [hunk 2]
```diff
@@ -444,7 +444,7 @@ SparseIntVect<boost::int64_t> *getTopologicalTorsionFingerprint(
         // look for a cycle that doesn't start at the first atom
         // we can't effectively canonicalize these at the moment
         // (was github #811)
-        if (pIt != path.begin() && pAtoms[*pIt]) {
+        if (pIt != path.begin() && *pIt != *(path.begin()) && pAtoms[*pIt]) {
```
Comment (bp-kelley) [inline Code/GraphMol/Fingerprints/AtomPairs.cpp]: Not blocking, but isn't this just?

```
if (*pIt != *(path.begin()) && pAtoms[*pIt])
```

I did the same thing internally when I looked at it so I also didn't notice the issue :)

Comment (bp-kelley) [inline Code/GraphMol/Fingerprints/AtomPairs.cpp]: I'll merge and see if the simplification works in another pull request.

**Pull request merged**